### PR TITLE
CLI client to run through demo workflow on migrating Coolstore

### PR DESCRIPTION
**Pull request by jwmatthews**
Related to #82 

A few things are addressed in this PR: (In hindsight I should have split things out into separate PRs)

- First cut at a client that parses analysis info for CoolStore and generates a requests per File with all incidents batched
- The response (received updated file) is saved and written to disk in the directory of the sample source code for Coolstore, so we are overwriting the exact file with the updates from Kai
- Tweaked logging in KAI Server to aid debugging and getting an idea of performance
- Added a Capture class to record data on server side to help us debug specific generated fixes

## To run this demo 
* Assumes you have the server running
1. `cd example`
2. `./fetch.sh`
3.  `source ../env/bin/activate`
4. `./run_demo.py`

Example of what the run_demo.py output looks like 
```
$ time ./run_demo.py
INFO - 2024-03-23 08:35:35,541 - [report.py:79 -         _read_report() ] - Reading report from ./analysis/coolstore/output.yaml
26 files with a total of 150 violations.
File #1 of 26 - Processing src/main/webapp/WEB-INF/web.xml which has 1 violations
Skipping src/main/webapp/WEB-INF/web.xml as it is not a Java file
File #2 of 26 - Processing pom.xml which has 14 violations
Skipping pom.xml as it is not a Java file
File #3 of 26 - Processing src/main/webapp/WEB-INF/beans.xml which has 5 violations
Skipping src/main/webapp/WEB-INF/beans.xml as it is not a Java file
File #4 of 26 - Processing src/main/resources/META-INF/persistence.xml which has 8 violations
Skipping src/main/resources/META-INF/persistence.xml as it is not a Java file
File #5 of 26 - Processing src/main/java/com/redhat/coolstore/model/InventoryEntity.java which has 6 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/model/InventoryEntity.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/model/InventoryEntity.java
File #6 of 26 - Processing src/main/java/com/redhat/coolstore/model/Order.java which has 9 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/model/Order.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/model/Order.java
File #7 of 26 - Processing src/main/java/com/redhat/coolstore/model/OrderItem.java which has 5 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/model/OrderItem.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/model/OrderItem.java
File #8 of 26 - Processing src/main/java/com/redhat/coolstore/model/ShoppingCart.java which has 1 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/model/ShoppingCart.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/model/ShoppingCart.java
File #9 of 26 - Processing src/main/java/com/redhat/coolstore/persistence/Resources.java which has 6 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/persistence/Resources.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/persistence/Resources.java
File #10 of 26 - Processing src/main/java/com/redhat/coolstore/rest/CartEndpoint.java which has 9 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/rest/CartEndpoint.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/rest/CartEndpoint.java
File #11 of 26 - Processing src/main/java/com/redhat/coolstore/rest/OrderEndpoint.java which has 8 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/rest/OrderEndpoint.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/rest/OrderEndpoint.java
File #12 of 26 - Processing src/main/java/com/redhat/coolstore/rest/ProductEndpoint.java which has 3 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/rest/ProductEndpoint.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/rest/ProductEndpoint.java
File #13 of 26 - Processing src/main/java/com/redhat/coolstore/rest/RestApplication.java which has 4 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/rest/RestApplication.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/rest/RestApplication.java
File #14 of 26 - Processing src/main/java/com/redhat/coolstore/service/CatalogService.java which has 8 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/service/CatalogService.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/service/CatalogService.java
File #15 of 26 - Processing src/main/java/com/redhat/coolstore/service/InventoryNotificationMDB.java which has 3 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/service/InventoryNotificationMDB.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/service/InventoryNotificationMDB.java
File #16 of 26 - Processing src/main/java/com/redhat/coolstore/service/OrderService.java which has 8 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/service/OrderService.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/service/OrderService.java
File #17 of 26 - Processing src/main/java/com/redhat/coolstore/service/OrderServiceMDB.java which has 15 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/service/OrderServiceMDB.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/service/OrderServiceMDB.java
File #18 of 26 - Processing src/main/java/com/redhat/coolstore/service/ProductService.java which has 3 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/service/ProductService.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/service/ProductService.java
File #19 of 26 - Processing src/main/java/com/redhat/coolstore/service/PromoService.java which has 1 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/service/PromoService.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/service/PromoService.java
File #20 of 26 - Processing src/main/java/com/redhat/coolstore/service/ShippingService.java which has 4 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/service/ShippingService.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/service/ShippingService.java
File #21 of 26 - Processing src/main/java/com/redhat/coolstore/service/ShoppingCartOrderProcessor.java which has 10 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/service/ShoppingCartOrderProcessor.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/service/ShoppingCartOrderProcessor.java
File #22 of 26 - Processing src/main/java/com/redhat/coolstore/service/ShoppingCartService.java which has 2 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/service/ShoppingCartService.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/service/ShoppingCartService.java
File #23 of 26 - Processing src/main/java/com/redhat/coolstore/utils/DataBaseMigrationStartup.java which has 7 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/utils/DataBaseMigrationStartup.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/utils/DataBaseMigrationStartup.java
File #24 of 26 - Processing src/main/java/com/redhat/coolstore/utils/Producers.java which has 3 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/utils/Producers.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/utils/Producers.java
File #25 of 26 - Processing src/main/java/com/redhat/coolstore/utils/StartupListener.java which has 1 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/utils/StartupListener.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/utils/StartupListener.java
File #26 of 26 - Processing src/main/java/com/redhat/coolstore/utils/Transformers.java which has 6 violations

Response StatusCode: 200 for src/main/java/com/redhat/coolstore/utils/Transformers.java

Writing to ./coolstore/src/main/java/com/redhat/coolstore/utils/Transformers.java
./run_demo.py  1.63s user 1.58s system 0% cpu 1:13:50.60 total

```

After running the demo we've updated the source code from coolstore git clone:

```
$ cd example/coolstore 
(env) [jmatthews] in ~/git/jwmatthews/kai/example/coolstore (main) $ git status
On branch main
Your branch is up to date with 'origin/main'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   src/main/java/com/redhat/coolstore/model/InventoryEntity.java
	modified:   src/main/java/com/redhat/coolstore/model/Order.java
	modified:   src/main/java/com/redhat/coolstore/model/OrderItem.java
	modified:   src/main/java/com/redhat/coolstore/model/ShoppingCart.java
	modified:   src/main/java/com/redhat/coolstore/persistence/Resources.java
	modified:   src/main/java/com/redhat/coolstore/rest/CartEndpoint.java
	modified:   src/main/java/com/redhat/coolstore/rest/OrderEndpoint.java
	modified:   src/main/java/com/redhat/coolstore/rest/ProductEndpoint.java
	modified:   src/main/java/com/redhat/coolstore/rest/RestApplication.java
	modified:   src/main/java/com/redhat/coolstore/service/CatalogService.java
	modified:   src/main/java/com/redhat/coolstore/service/InventoryNotificationMDB.java
	modified:   src/main/java/com/redhat/coolstore/service/OrderService.java
	modified:   src/main/java/com/redhat/coolstore/service/OrderServiceMDB.java
	modified:   src/main/java/com/redhat/coolstore/service/ProductService.java
	modified:   src/main/java/com/redhat/coolstore/service/PromoService.java
	modified:   src/main/java/com/redhat/coolstore/service/ShippingService.java
	modified:   src/main/java/com/redhat/coolstore/service/ShoppingCartOrderProcessor.java
	modified:   src/main/java/com/redhat/coolstore/service/ShoppingCartService.java
	modified:   src/main/java/com/redhat/coolstore/utils/DataBaseMigrationStartup.java
	modified:   src/main/java/com/redhat/coolstore/utils/Producers.java
	modified:   src/main/java/com/redhat/coolstore/utils/StartupListener.java
	modified:   src/main/java/com/redhat/coolstore/utils/Transformers.java

no changes added to commit (use "git add" and/or "git commit -a")
```

Below is a summary of what we've changed in the Coolstore app
```
---
 src/main/java/com/redhat/coolstore/model/InventoryEntity.java              |  77 +++++++++++-----------
 src/main/java/com/redhat/coolstore/model/Order.java                        | 256 ++++++++++++++++++++++++++++++++++++++++++++----------------------------
 src/main/java/com/redhat/coolstore/model/OrderItem.java                    | 145 +++++++++++++++++++++++++++++++----------
 src/main/java/com/redhat/coolstore/model/ShoppingCart.java                 | 200 +++++++++++++++++++++++++++-----------------------------
 src/main/java/com/redhat/coolstore/persistence/Resources.java              |  33 ++++++----
 src/main/java/com/redhat/coolstore/rest/CartEndpoint.java                  | 160 +++++----------------------------------------
 src/main/java/com/redhat/coolstore/rest/OrderEndpoint.java                 |  54 ++++++++++-----
 src/main/java/com/redhat/coolstore/rest/ProductEndpoint.java               |  17 +++--
 src/main/java/com/redhat/coolstore/rest/RestApplication.java               |  50 +++++++++++++-
 src/main/java/com/redhat/coolstore/service/CatalogService.java             |  51 +++------------
 src/main/java/com/redhat/coolstore/service/InventoryNotificationMDB.java   |  66 +++++--------------
 src/main/java/com/redhat/coolstore/service/OrderService.java               |  40 +++---------
 src/main/java/com/redhat/coolstore/service/OrderServiceMDB.java            |  77 +++++++++++-----------
 src/main/java/com/redhat/coolstore/service/ProductService.java             |  40 +++++++-----
 src/main/java/com/redhat/coolstore/service/PromoService.java               | 102 +++++++++--------------------
 src/main/java/com/redhat/coolstore/service/ShippingService.java            |  82 +++--------------------
 src/main/java/com/redhat/coolstore/service/ShoppingCartOrderProcessor.java |  47 +++++---------
 src/main/java/com/redhat/coolstore/service/ShoppingCartService.java        |  75 ++++++---------------
 src/main/java/com/redhat/coolstore/utils/DataBaseMigrationStartup.java     |  76 ++++++++++------------
 src/main/java/com/redhat/coolstore/utils/Producers.java                    |  21 +++---
 src/main/java/com/redhat/coolstore/utils/StartupListener.java              |  20 ++++--
 src/main/java/com/redhat/coolstore/utils/Transformers.java                 |  57 ++++++++--------
 22 files changed, 792 insertions(+), 954 deletions(-)
```

Example of what the logging output on server side looks like now with changes in this PR

```
$ make run-server
PYTHONPATH="/Users/jmatthews/git/jwmatthews/kai/kai:" python ./kai/server.py --loglevel "info"
Logging for KAI has been initialized and the level set to INFO
INFO - 2024-03-23 07:53:13,794 - [server.py:482 -             <module>() ] - Config loaded: {'embeddings': {'todo': True},
 'models': {'args': {'model_id': 'ibm/granite-13b-chat-v2'},
            'provider': 'IBMGranite'},
 'postgresql': {'database': 'kai',
                'host': '127.0.0.1',
                'password': 'dog8code',
                'user': 'kai'}}
INFO - 2024-03-23 07:53:13,818 - [incident_store_advanced.py:121 -             __init__() ] - Connected to the PostgreSQL server.
INFO - 2024-03-23 07:53:13,824 - [server.py:506 -             <module>() ] - Selected model IBMGranite
======== Running on http://0.0.0.0:8080 ========
(Press CTRL+C to quit)
INFO - 2024-03-23 07:53:15,602 - [server.py:412 - get_incident_solutions_for_file() ] - START - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' with 6 incidents'
INFO - 2024-03-23 07:53:15,602 - [server.py:428 - get_incident_solutions_for_file() ] - Processing incident 1/6 for src/main/java/com/redhat/coolstore/model/InventoryEntity.java
INFO - 2024-03-23 07:53:15,602 - [server.py:239 - get_incident_solution() ] - START - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '5' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:53:52,572 - [server.py:292 - get_incident_solution() ] - END - completed in '36.97047185897827s: - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '5' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:53:52,573 - [server.py:428 - get_incident_solutions_for_file() ] - Processing incident 2/6 for src/main/java/com/redhat/coolstore/model/InventoryEntity.java
INFO - 2024-03-23 07:53:52,573 - [server.py:239 - get_incident_solution() ] - START - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '6' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:54:29,924 - [server.py:292 - get_incident_solution() ] - END - completed in '37.35103416442871s: - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '6' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:54:29,924 - [server.py:428 - get_incident_solutions_for_file() ] - Processing incident 3/6 for src/main/java/com/redhat/coolstore/model/InventoryEntity.java
INFO - 2024-03-23 07:54:29,924 - [server.py:239 - get_incident_solution() ] - START - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '7' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:55:06,836 - [server.py:292 - get_incident_solution() ] - END - completed in '36.91150522232056s: - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '7' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:55:06,836 - [server.py:428 - get_incident_solutions_for_file() ] - Processing incident 4/6 for src/main/java/com/redhat/coolstore/model/InventoryEntity.java
INFO - 2024-03-23 07:55:06,836 - [server.py:239 - get_incident_solution() ] - START - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '8' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:55:44,201 - [server.py:292 - get_incident_solution() ] - END - completed in '37.36531090736389s: - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '8' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:55:44,201 - [server.py:428 - get_incident_solutions_for_file() ] - Processing incident 5/6 for src/main/java/com/redhat/coolstore/model/InventoryEntity.java
INFO - 2024-03-23 07:55:44,201 - [server.py:239 - get_incident_solution() ] - START - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '9' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:56:20,819 - [server.py:292 - get_incident_solution() ] - END - completed in '36.61733293533325s: - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '9' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:56:20,819 - [server.py:428 - get_incident_solutions_for_file() ] - Processing incident 6/6 for src/main/java/com/redhat/coolstore/model/InventoryEntity.java
INFO - 2024-03-23 07:56:20,819 - [server.py:239 - get_incident_solution() ] - START - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '10' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:56:53,665 - [server.py:292 - get_incident_solution() ] - END - completed in '32.8460590839386s: - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '10' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:56:53,665 - [server.py:438 - get_incident_solutions_for_file() ] - END - completed in '218.06388401985168s:  - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' with 6 incidents'
```

This is an example of what the 'capture' class is doing, it's gathering related info and writing to disk from server side to help us with debugging specific generated fixes.

```
$ tree logs
logs
└── ibm
    └── granite-13b-chat-v2
        └── coolstore
            ├── src_main_java_com_redhat_coolstore_model_InventoryEntity.java
            │   └── eap8
            │       └── eap7
            │           └── javax-to-jakarta-import-00001
            │               ├── llm_result
            │               ├── prompt
            │               ├── request.json
            │               ├── solved_incident.json
            │               └── src_main_java_com_redhat_coolstore_model_InventoryEntity.java
            └── src_main_java_com_redhat_coolstore_model_Order.java
                └── eap8
                    └── eap7
                        └── javax-to-jakarta-import-00001
                            ├── llm_result
                            ├── prompt
                            ├── request.json
                            ├── solved_incident.json
                            └── src_main_java_com_redhat_coolstore_model_Order.java

12 directories, 10 files
```

### CODE REVIEW

File: kai/report.py [hunk 1]
```diff
@@ -5,7 +5,8 @@
 from io import StringIO
 
 import yaml
-from kai_logging import KAI_LOG
+
+from kai.kai_logging import KAI_LOG
```
Comment (jwmatthews) [inline kai/report.py:9]: I needed to tweak a few of the kai_logging imports, was seeing below when trying to run


```
$ time ./run_demo.py           
Traceback (most recent call last):
  File "/Users/jmatthews/git/jwmatthews/kai/example/./run_demo.py", line 12, in <module>
    from kai import Report
  File "/Users/jmatthews/git/jwmatthews/kai/example/../../kai/kai/__init__.py", line 7, in <module>
    from kai.incident_store import IncidentStore
  File "/Users/jmatthews/git/jwmatthews/kai/example/../../kai/kai/incident_store.py", line 11, in <module>
    from kai.report import Report
  File "/Users/jmatthews/git/jwmatthews/kai/example/../../kai/kai/report.py", line 8, in <module>
    from kai_logging import KAI_LOG
ModuleNotFoundError: No module named 'kai_logging'
```

**Comment by jwmatthews** [comment]
Example of what the logging output on server side looks like now with changes in this PR

```
$ make run-server
PYTHONPATH="/Users/jmatthews/git/jwmatthews/kai/kai:" python ./kai/server.py --loglevel "info"
Logging for KAI has been initialized and the level set to INFO
INFO - 2024-03-23 07:53:13,794 - [server.py:482 -             <module>() ] - Config loaded: {'embeddings': {'todo': True},
 'models': {'args': {'model_id': 'ibm/granite-13b-chat-v2'},
            'provider': 'IBMGranite'},
 'postgresql': {'database': 'kai',
                'host': '127.0.0.1',
                'password': 'dog8code',
                'user': 'kai'}}
INFO - 2024-03-23 07:53:13,818 - [incident_store_advanced.py:121 -             __init__() ] - Connected to the PostgreSQL server.
INFO - 2024-03-23 07:53:13,824 - [server.py:506 -             <module>() ] - Selected model IBMGranite
======== Running on http://0.0.0.0:8080 ========
(Press CTRL+C to quit)
INFO - 2024-03-23 07:53:15,602 - [server.py:412 - get_incident_solutions_for_file() ] - START - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' with 6 incidents'
INFO - 2024-03-23 07:53:15,602 - [server.py:428 - get_incident_solutions_for_file() ] - Processing incident 1/6 for src/main/java/com/redhat/coolstore/model/InventoryEntity.java
INFO - 2024-03-23 07:53:15,602 - [server.py:239 - get_incident_solution() ] - START - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '5' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:53:52,572 - [server.py:292 - get_incident_solution() ] - END - completed in '36.97047185897827s: - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '5' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:53:52,573 - [server.py:428 - get_incident_solutions_for_file() ] - Processing incident 2/6 for src/main/java/com/redhat/coolstore/model/InventoryEntity.java
INFO - 2024-03-23 07:53:52,573 - [server.py:239 - get_incident_solution() ] - START - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '6' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:54:29,924 - [server.py:292 - get_incident_solution() ] - END - completed in '37.35103416442871s: - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '6' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:54:29,924 - [server.py:428 - get_incident_solutions_for_file() ] - Processing incident 3/6 for src/main/java/com/redhat/coolstore/model/InventoryEntity.java
INFO - 2024-03-23 07:54:29,924 - [server.py:239 - get_incident_solution() ] - START - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '7' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:55:06,836 - [server.py:292 - get_incident_solution() ] - END - completed in '36.91150522232056s: - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '7' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:55:06,836 - [server.py:428 - get_incident_solutions_for_file() ] - Processing incident 4/6 for src/main/java/com/redhat/coolstore/model/InventoryEntity.java
INFO - 2024-03-23 07:55:06,836 - [server.py:239 - get_incident_solution() ] - START - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '8' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:55:44,201 - [server.py:292 - get_incident_solution() ] - END - completed in '37.36531090736389s: - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '8' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:55:44,201 - [server.py:428 - get_incident_solutions_for_file() ] - Processing incident 5/6 for src/main/java/com/redhat/coolstore/model/InventoryEntity.java
INFO - 2024-03-23 07:55:44,201 - [server.py:239 - get_incident_solution() ] - START - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '9' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:56:20,819 - [server.py:292 - get_incident_solution() ] - END - completed in '36.61733293533325s: - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '9' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:56:20,819 - [server.py:428 - get_incident_solutions_for_file() ] - Processing incident 6/6 for src/main/java/com/redhat/coolstore/model/InventoryEntity.java
INFO - 2024-03-23 07:56:20,819 - [server.py:239 - get_incident_solution() ] - START - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '10' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:56:53,665 - [server.py:292 - get_incident_solution() ] - END - completed in '32.8460590839386s: - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' 'eap8/eap7'/'javax-to-jakarta-import-00001' @ Line Number '10' using model_id 'ibm/granite-13b-chat-v2'
INFO - 2024-03-23 07:56:53,665 - [server.py:438 - get_incident_solutions_for_file() ] - END - completed in '218.06388401985168s:  - App: 'coolstore', File: 'src/main/java/com/redhat/coolstore/model/InventoryEntity.java' with 6 incidents'
```

**Comment by jwmatthews** [comment]
This is an example of what the 'capture' class is doing, it's gathering related info and writing to disk from server side to help us with debugging specific generated fixes.

```
$ tree logs
logs
└── ibm
    └── granite-13b-chat-v2
        └── coolstore
            ├── src_main_java_com_redhat_coolstore_model_InventoryEntity.java
            │   └── eap8
            │       └── eap7
            │           └── javax-to-jakarta-import-00001
            │               ├── llm_result
            │               ├── prompt
            │               ├── request.json
            │               ├── solved_incident.json
            │               └── src_main_java_com_redhat_coolstore_model_InventoryEntity.java
            └── src_main_java_com_redhat_coolstore_model_Order.java
                └── eap8
                    └── eap7
                        └── javax-to-jakarta-import-00001
                            ├── llm_result
                            ├── prompt
                            ├── request.json
                            ├── solved_incident.json
                            └── src_main_java_com_redhat_coolstore_model_Order.java

12 directories, 10 files
```

**Comment by jwmatthews** [comment]
@JonahSussman @jmontleon @savitharaghunathan @fabianvf this is ready to review when anyone has time.

I would ask that for purposes of this PR we just consider the ability to kick off a workflow to migrate Coolstore via KAI.  The quality of the results I believe will need to be refined in subsequent PRs.

File: Makefile [hunk 2]
```diff
@@ -7,7 +7,10 @@ run-postgres:
 	$(CONTAINER_RUNTIME) run -it -v data:/var/lib/postgresql/data -e POSTGRES_USER=kai -e POSTGRES_PASSWORD=dog8code -e POSTGRES_DB=kai -p 5432:5432 docker.io/pgvector/pgvector:pg15
 
 run-server:
-	PYTHONPATH=$(KAI_PYTHON_PATH) python ./kai/server.py
+	PYTHONPATH=$(KAI_PYTHON_PATH) python ./kai/server.py --loglevel "info"
+
+run-server-debug:
+	PYTHONPATH=$(KAI_PYTHON_PATH) python ./kai/server.py --loglevel "debug"
```
[the remark this answers is not in the document]
Comment (jwmatthews) [reply Makefile:13]: Thank you for suggestion, will update.

File: example/README.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# To run
+
+- Assumes that you have `make run-server` running from base directory
+
+1. `fetch.sh`
+1. `source ../env/bin/activate`
```
[the remark this answers is not in the document]
Comment (jwmatthews) [reply example/README.md:6]: I was assuming we want to recommend that the reader activates the same virtual env used for the kai server.

[the remark this answers is not in the document]
Comment (fabianvf) [reply example/README.md:6]: that makes sense, but I don't use same virtualenv setup so I don't have a `../env` directory. Might make more sense to just add a note making sure you use the same python environment

File: Makefile [hunk 4]
```diff
@@ -7,7 +7,10 @@ run-postgres:
...
 
 load-data:
 	PYTHONPATH=$(KAI_PYTHON_PATH) python ./kai/incident_store_advanced.py --config_filepath ./kai/database.ini
```
Comment (savitharaghunathan) [inline Makefile:16]: Looks the database.ini doesnt exist anymore. change to 
```
PYTHONPATH=$(KAI_PYTHON_PATH) python ./kai/incident_store_advanced.py --config_filepath ./kai/config.toml

Comment (savitharaghunathan) [inline Makefile:16]: Looks the database.ini doesnt exist anymore. change to 
```
PYTHONPATH=$(KAI_PYTHON_PATH) python ./kai/incident_store_advanced.py --config_filepath ./kai/config.toml

Comment (jwmatthews) [reply Makefile:16]: Thank you @savitharaghunathan for catching that!

**Comment by jwmatthews** [comment]
Updated and addressed suggestions

**Review by JonahSussman — APPROVED** [review]
+1 to the server logging

**Review by savitharaghunathan — APPROVED** [review]

**Pull request merged**